exp/api/v1beta1/azuremanagedmachinepool_webhook.go

```diff
@@ -131,6 +133,25 @@ func (r *AzureManagedMachinePool) ValidateUpdate(oldRaw runtime.Object, client c
 		}
 	}
 
+	if old.Spec.OsDiskType != "" {
+		// Prevent OSDiskType modification if it was already set to some value
+		if r.Spec.OsDiskType == "" {
+			// unsetting the field is not allowed
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("Spec", "OsDiskType"),
+					r.Spec.OsDiskType,
+					"field is immutable, unsetting is not allowed"))
+		} else if r.Spec.OsDiskType != old.Spec.OsDiskType {
+			// changing the field is not allowed
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("Spec", "OsDiskType"),
+					r.Spec.OsDiskType,
+					"field is immutable"))
+		}
+	}
+
```

Comment:
Do we need to handle both situations individually? Can they not be combined into a single "immutable" error?

```suggestion
	if old.Spec.OsDiskType != "" && r.Spec.OsDiskType != old.Spec.OsDiskType  {
		// Prevent OSDiskType modification if it was already set to some value
		allErrs = append(allErrs,
			field.Invalid(
				field.NewPath("Spec", "OsDiskType"),
				r.Spec.OsDiskType,
				"field is immutable"))
	}

```

Reply:
I think your observation here is sound. If `A` is a non-empty string, then we can simply test `B` for being different from `A`. `B` being an empty string will always be different from `A` when `A` is a non-empty string, so we can exclusively use:

- `if r.Spec.OsDiskType != old.Spec.OsDiskType`

Reply:
Because other validations in this surface area are doing this "test for empty string" validation, let's continue this pattern for now. If we want to change this approach let's get consensus and then change it everywhere so that we're consistent.

Thoughts?

Reply:
+1 for consistency. I hadn't noticed the similar pattern above.